File: node/src/sensor/heat.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of a heat sensor.
 *
 * @internal
 */
export enum HeatProperty {
  /** Whether abnormal heat is detected. */
  Detected = 'detected',
}

/**
 * Property values of a heat sensor.
 *
 * @internal
 */
export interface HeatSensorProperties {
  [HeatProperty.Detected]: boolean;
}

/** Read-only proxy interface for a heat sensor. */
export interface HeatSensorLike extends SensorLike {
  readonly type: SensorType.Heat;
  readonly onPropertyChanged: Observable<PropertyChangeOf<HeatSensorProperties>>;

  getValue(property: HeatProperty.Detected): boolean | undefined;
  getValue(property: string): unknown;
}

/** Heat alarm sensor. */
export class HeatSensor<TStorage extends object = Record<string, any>> extends Sensor<HeatSensorProperties, TStorage> {
  readonly type = SensorType.Heat;
  readonly category = SensorCategory.Sensor;

  constructor(name = 'Heat Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [HeatProperty.Detected]: false });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  /**
   * Report abnormal heat detection state.
   *
   * @param value - True when heat is currently detected.
   *
   * @example
   * ```ts
   * heat.setDetected(true);
   * ```
   */
  setDetected(value: boolean): void {
    this._writeState({ [HeatProperty.Detected]: value });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link HeatSensor}. */
export const heatMeta = defineSensor({
  type: SensorType.Heat,
  category: SensorCategory.Sensor,
  assignmentKey: 'heat',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [HeatProperty.Detected]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: HeatProperty.Detected, value: true, sustained: true },
  virtual: { properties: { [HeatProperty.Detected]: false } },
  semantics: {
    domain: SensorDomain.Binary,
    stateProperty: HeatProperty.Detected,
    commandProperty: HeatProperty.Detected,
    deviceClass: 'heat',
  },
});

File: python/camera_ui_sdk/sensor/heat.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class HeatProperty(StrEnum):
    """Property names of a heat sensor."""

    Detected = "detected"
    """Whether abnormal heat is detected."""


class HeatSensorProperties(TypedDict):
    """Property values of a heat sensor."""

    detected: bool


class HeatPropertyChangeData(TypedDict):
    """Property change payload emitted on HeatSensorLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a HeatProperty value."""
    value: bool
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class HeatSensorLike(SensorLike, Protocol):
    """Read-only proxy interface for a heat sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Heat

    @property
    def onPropertyChanged(self) -> Observable[HeatPropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[HeatProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class HeatSensor(Sensor[HeatSensorProperties, TStorage, str], Generic[TStorage]):
    """Heat alarm sensor."""

    _requires_frames = False

    def __init__(self, name: str = "Heat Sensor", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state({HeatProperty.Detected.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Heat

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        return bool(self.props.detected)

    def setDetected(self, value: bool) -> None:
        """Report abnormal heat detection state.

        Args:
            value: True when heat is currently detected.

        Example:
            ```python
            heat.setDetected(True)
            ```
        """
        self._write_state({HeatProperty.Detected.value: value})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""
